lib/hashie/extensions/strict_key_access.rb

```diff
@@ -0,0 +1,71 @@
+module Hashie
+  module Extensions
+    # SRP: This extension will fail an error whenever a key is accessed that does not exist in the hash.
+    #
+    #   EXAMPLE:
+    #
+    #     class StrictKeyAccessHash < Hash
+    #       include Hashie::Extensions::StrictKeyAccess
+    #     end
+    #
+    #     >> hash = StrictKeyAccessHash[foo: "bar"]
+    #     => {:foo=>"bar"}
+    #     >> hash[:foo]
+    #     => "bar"
+    #     >> hash[:cow]
+    #       KeyError: key not found: :cow
+    #
+    # NOTE: For googlers coming from Python to Ruby, this extension makes a Hash behave like a "Dictionary".
+    #
+    module StrictKeyAccess
+      class DefaultError < StandardError
+        def initialize(msg = 'Setting or using a default with Hashie::Extensions::StrictKeyAccess does not make sense', *args)
+          super
+        end
+      end
+
+      # NOTE: This extension would break the default behavior of Hash initialization:
+      #
+      #     >> a = StrictKeyAccessHash.new(a: :b)
+      #     => {}
+      #     >> a[:a]
+      #       KeyError: key not found: :a
+      #
+      # Includes the Hashie::Extensions::MergeInitializer extension to get around that problem.
+      # Also note that defaults still don't make any sense with a StrictKeyAccess.
+      def self.included(base)
+        # Can only include into classes with a hash initializer
+        base.send(:include, Hashie::Extensions::MergeInitializer)
```

Comment:
@michaelherold See the note at the top of the comment above:

> # NOTE: This extension would break the default behavior of Hash initialization:

Here's a demo:

```
>> class BasicStrictHash < Hash
>> def [](key); fetch(key); end
>> end
=> :[]
>> basic_strict_hash = BasicStrictHash[a: :b] # works!
=> {:a=>:b}
>> basic_strict_hash_via_new = BasicStrictHash.new(a: :b) # does *not* work!
=> {}
>> basic_strict_hash[:a]
=> :b
>> basic_strict_hash_via_new[:a] # ERROR!!!
KeyError: key not found: :a
	from (irb):2:in `fetch'
	from (irb):2:in `[]'
	from (irb):7
	from bin/console:14:in `<main>'
```
When we use the MergeInitializer... magic happens:
```
>> class StrictHashWithFixedInitializer < Hash
>> include Hashie::Extensions::MergeInitializer
>> def [](key); fetch(key); end
>> end
=> :[]
>> basic_strict_hash = BasicStrictHash[a: :b] # still works!
=> {:a=>:b}
>> strict_hash_with_fixed_initializer = StrictHashWithFixedInitializer.new(a: :b) # now it works!
=> {:a=>:b}
>> strict_hash_with_fixed_initializer[:a]
=> :b
```

Does that make it clear?